File: web/src/api/scorecard.ts
/**
 * API client for Validation Scorecard endpoints.
 */
import client from './client';

// ============================================================================
// Types
// ============================================================================

/**
 * Valid rating values for scorecard criteria.
 */
export type ScorecardRating = 'Green' | 'Green-' | 'Yellow+' | 'Yellow' | 'Yellow-' | 'Red' | 'N/A' | null;

/**
 * Rating options for dropdowns with display labels.
 */
export const RATING_OPTIONS: { value: ScorecardRating; label: string; color: string }[] = [
    { value: null, label: 'Select Rating...', color: 'gray' },
    { value: 'Green', label: 'Green (6)', color: 'green' },
    { value: 'Green-', label: 'Green- (5)', color: 'green' },
    { value: 'Yellow+', label: 'Yellow+ (4)', color: 'yellow' },
    { value: 'Yellow', label: 'Yellow (3)', color: 'yellow' },
    { value: 'Yellow-', label: 'Yellow- (2)', color: 'yellow' },
    { value: 'Red', label: 'Red (1)', color: 'red' },
    { value: 'N/A', label: 'N/A (0)', color: 'gray' },
];

/**
 * Scorecard criterion from configuration.
 */
export interface ScorecardCriterion {
    criterion_id: number;
    code: string;
    section_id: number;
    name: string;
    description_prompt: string | null;
    comments_prompt: string | null;
    include_in_summary: boolean;
    allow_zero: boolean;
    weight: number;
    sort_order: number;
    is_active: boolean;
    created_at: string;
    updated_at: string;
}

/**
 * Scorecard section with nested criteria.
 */
export interface ScorecardSection {
    section_id: number;
    code: string;
    name: string;
    description: string | null;
    sort_order: number;
    is_active: boolean;
    created_at: string;
    updated_at: string;
    criteria: ScorecardCriterion[];
}

/**
 * Scorecard configuration response.
 */
export interface ScorecardConfigResponse {
    sections: ScorecardSection[];
}

/**
 * Input for a single criterion rating.
 */
export interface CriterionRatingInput {
    criterion_code: string;
    rating: ScorecardRating;
    description?: string | null;
    comments?: string | null;
}

/**
 * Input for updating a single criterion rating.
 */
export interface CriterionRatingUpdate {
    rating?: ScorecardRating;
    description?: string | null;
    comments?: string | null;
}

/**
 * Request body for creating/updating all ratings.
 */
export interface ScorecardRatingsCreate {
    ratings: CriterionRatingInput[];
}

/**
 * Criterion detail in scorecard response.
 */
export interface CriterionDetailResponse {
    criterion_code: string;
    criterion_name: string;
    section_code: string;
    rating: ScorecardRating;
    numeric_score: number;
    description: string | null;
    comments: string | null;
}

/**
 * Section summary in scorecard response.
 */
export interface SectionSummaryResponse {
    section_code: string;
    section_name: string;
    criteria_count: number;
    rated_count: number;
    unrated_count: number;
    numeric_score: number;
    rating: string | null;
}

/**
 * Overall assessment in scorecard response.
 */
export interface OverallAssessmentResponse {
    numeric_score: number;
    rating: string | null;
    sections_count: number;
    rated_sections_count: number;
    overall_assessment_narrative: string | null;
}

/**
 * Full scorecard response with all details.
 */
export interface ScorecardFullResponse {
    request_id: number;
    criteria_details: CriterionDetailResponse[];
    section_summaries: SectionSummaryResponse[];
    overall_assessment: OverallAssessmentResponse;
    computed_at: string;
}

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Convert a rating to its display color class.
 */
export function getRatingColorClass(rating: string | null): string {
    if (!rating) return 'bg-gray-100 text-gray-600';

    switch (rating) {
        case 'Green':
        case 'Green-':
            return 'bg-green-100 text-green-800';
        case 'Yellow+':
        case 'Yellow':
        case 'Yellow-':
            return 'bg-yellow-100 text-yellow-800';
        case 'Red':
            return 'bg-red-100 text-red-800';
        case 'N/A':
            return 'bg-gray-100 text-gray-600';
        default:
            return 'bg-gray-100 text-gray-600';
    }
}

/**
 * Convert a numeric score to its display color class.
 */
export function getScoreColorClass(score: number): string {
    if (score >= 5) return 'text-green-600';
    if (score >= 3) return 'text-yellow-600';
    if (score >= 1) return 'text-red-600';
    return 'text-gray-400';
}

// ============================================================================
// API Functions
// ============================================================================

const BASE_URL = '/scorecard';

/**
 * Get the scorecard configuration (sections and criteria).
 */
export async function getScorecardConfig(): Promise<ScorecardConfigResponse> {
    const response = await client.get(`${BASE_URL}/config`);
    return response.data;
}

/**
 * Get the scorecard for a validation request.
 */
export async function getScorecard(requestId: number): Promise<ScorecardFullResponse> {
    const response = await client.get(`${BASE_URL}/validation/${requestId}`);
    return response.data;
}

/**
 * Create or update all scorecard ratings for a validation request.
 */
export async function saveScorecard(
    requestId: number,
    ratings: CriterionRatingInput[]
): Promise<ScorecardFullResponse> {
    const response = await client.post(`${BASE_URL}/validation/${requestId}`, {
        ratings
    });
    return response.data;
}

/**
 * Update a single criterion rating.
 */
export async function updateSingleRating(
    requestId: number,
    criterionCode: string,
    data: CriterionRatingUpdate
): Promise<ScorecardFullResponse> {
    const response = await client.patch(
        `${BASE_URL}/validation/${requestId}/ratings/${criterionCode}`,
        data
    );
    return response.data;
}

/**
 * Update the overall assessment narrative.
 */
export async function updateOverallNarrative(
    requestId: number,
    narrative: string | null
): Promise<ScorecardFullResponse> {
    const response = await client.patch(
        `${BASE_URL}/validation/${requestId}/overall-narrative`,
        { overall_assessment_narrative: narrative }
    );
    return response.data;
}

/**
 * Export the validation scorecard as a PDF.
 * Returns a Blob that can be downloaded by the browser.
 */
export async function exportScorecardPDF(requestId: number): Promise<Blob> {
    const response = await client.get(
        `${BASE_URL}/validation/${requestId}/export-pdf`,
        { responseType: 'blob' }
    );
    return response.data;
}

// ============================================================================
// Admin: Section CRUD Functions
// ============================================================================

/**
 * Input for creating a new section.
 */
export interface ScorecardSectionCreate {
    code: string;
    name: string;
    description?: string | null;
    sort_order?: number;
    is_active?: boolean;
}

/**
 * Input for updating a section.
 */
export interface ScorecardSectionUpdate {
    name?: string;
    description?: string | null;
    sort_order?: number;
    is_active?: boolean;
}

/**
 * List all scorecard sections.
 */
export async function listSections(includeInactive = false): Promise<ScorecardSection[]> {
    const response = await client.get(`${BASE_URL}/sections`, {
        params: { include_inactive: includeInactive }
    });
    return response.data;
}

/**
 * Get a single section with its criteria.
 */
export async function getSection(sectionId: number): Promise<ScorecardSection> {
    const response = await client.get(`${BASE_URL}/sections/${sectionId}`);
    return response.data;
}

/**
 * Create a new section.
 */
export async function createSection(data: ScorecardSectionCreate): Promise<ScorecardSection> {
    const response = await client.post(`${BASE_URL}/sections`, data);
    return response.data;
}

/**
 * Update a section.
 */
export async function updateSection(sectionId: number, data: ScorecardSectionUpdate): Promise<ScorecardSection> {
    const response = await client.patch(`${BASE_URL}/sections/${sectionId}`, data);
    return response.data;
}

/**
 * Delete a section.
 */
export async function deleteSection(sectionId: number): Promise<void> {
    await client.delete(`${BASE_URL}/sections/${sectionId}`);
}

// ============================================================================
// Admin: Criterion CRUD Functions
// ============================================================================

/**
 * Input for creating a new criterion.
 */
export interface ScorecardCriterionCreate {
    code: string;
    section_id: number;
    name: string;
    description_prompt?: string | null;
    comments_prompt?: string | null;
    include_in_summary?: boolean;
    allow_zero?: boolean;
    weight?: number;
    sort_order?: number;
    is_active?: boolean;
}

/**
 * Input for updating a criterion.
 */
export interface ScorecardCriterionUpdate {
    name?: string;
    description_prompt?: string | null;
    comments_prompt?: string | null;
    include_in_summary?: boolean;
    allow_zero?: boolean;
    weight?: number;
    sort_order?: number;
    is_active?: boolean;
}

/**
 * List all scorecard criteria.
 */
export async function listCriteria(sectionId?: number, includeInactive = false): Promise<ScorecardCriterion[]> {
    const response = await client.get(`${BASE_URL}/criteria`, {
        params: {
            section_id: sectionId,
            include_inactive: includeInactive
        }
    });
    return response.data;
}

/**
 * Get a single criterion.
 */
export async function getCriterion(criterionId: number): Promise<ScorecardCriterion> {
    const response = await client.get(`${BASE_URL}/criteria/${criterionId}`);
    return response.data;
}

/**
 * Create a new criterion.
 */
export async function createCriterion(data: ScorecardCriterionCreate): Promise<ScorecardCriterion> {
    const response = await client.post(`${BASE_URL}/criteria`, data);
    return response.data;
}

/**
 * Update a criterion.
 */
export async function updateCriterion(criterionId: number, data: ScorecardCriterionUpdate): Promise<ScorecardCriterion> {
    const response = await client.patch(`${BASE_URL}/criteria/${criterionId}`, data);
    return response.data;
}

/**
 * Delete a criterion.
 */
export async function deleteCriterion(criterionId: number): Promise<void> {
    await client.delete(`${BASE_URL}/criteria/${criterionId}`);
}

// ============================================================================
// Configuration Versioning Types
// ============================================================================

/**
 * Section snapshot in a config version.
 */
export interface ScorecardSectionSnapshot {
    snapshot_id: number;
    code: string;
    name: string;
    description: string | null;
    sort_order: number;
    is_active: boolean;
}

/**
 * Criterion snapshot in a config version.
 */
export interface ScorecardCriterionSnapshot {
    snapshot_id: number;
    section_code: string;
    code: string;
    name: string;
    description_prompt: string | null;
    comments_prompt: string | null;
    include_in_summary: boolean;
    allow_zero: boolean;
    weight: number;
    sort_order: number;
    is_active: boolean;
}

/**
 * Scorecard configuration version summary.
 */
export interface ScorecardConfigVersion {
    version_id: number;
    version_number: number;
    version_name: string | null;
    description: string | null;
    published_by_name: string | null;
    published_at: string;
    is_active: boolean;
    sections_count: number;
    criteria_count: number;
    scorecards_count: number;
    created_at: string;
    has_unpublished_changes: boolean;
}

/**
 * Scorecard configuration version with full snapshots.
 */
export interface ScorecardConfigVersionDetail extends ScorecardConfigVersion {
    section_snapshots: ScorecardSectionSnapshot[];
    criterion_snapshots: ScorecardCriterionSnapshot[];
}

/**
 * Input for publishing a new version.
 */
export interface PublishScorecardVersionRequest {
    version_name?: string | null;
    description?: string | null;
}

// ============================================================================
// Configuration Versioning API Functions
// ============================================================================

/**
 * List all scorecard configuration versions.
 */
export async function listConfigVersions(): Promise<ScorecardConfigVersion[]> {
    const response = await client.get(`${BASE_URL}/versions`);
    return response.data;
}

/**
 * Get the currently active scorecard configuration version with full details.
 */
export async function getActiveConfigVersion(): Promise<ScorecardConfigVersionDetail | null> {
    const response = await client.get(`${BASE_URL}/versions/active`);
    return response.data;
}

/**
 * Get a specific scorecard configuration version with full details.
 */
export async function getConfigVersion(versionId: number): Promise<ScorecardConfigVersionDetail> {
    const response = await client.get(`${BASE_URL}/versions/${versionId}`);
    return response.data;
}

/**
 * Publish a new scorecard configuration version. Admin only.
 */
export async function publishConfigVersion(data: PublishScorecardVersionRequest): Promise<ScorecardConfigVersion> {
    const response = await client.post(`${BASE_URL}/versions/publish`, data);
    return response.data;
}

export default {
    getScorecardConfig,
    getScorecard,
    saveScorecard,
    updateSingleRating,
    updateOverallNarrative,
    exportScorecardPDF,
    getRatingColorClass,
    getScoreColorClass,
    RATING_OPTIONS,
    // Admin functions
    listSections,
    getSection,
    createSection,
    updateSection,
    deleteSection,
    listCriteria,
    getCriterion,
    createCriterion,
    updateCriterion,
    deleteCriterion,
    // Config versioning
    listConfigVersions,
    getActiveConfigVersion,
    getConfigVersion,
    publishConfigVersion,
};

File: api/app/schemas/scorecard.py
"""Pydantic schemas for Validation Scorecard API."""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


# ============================================================================
# Valid Rating Values
# ============================================================================

VALID_RATINGS = ["Green", "Green-", "Yellow+",
                 "Yellow", "Yellow-", "Red", "N/A"]
RATING_PATTERN = "^(Green|Green-|Yellow\\+|Yellow|Yellow-|Red|N/A)$"


# ============================================================================
# Configuration Schemas (Sections & Criteria)
# ============================================================================

class ScorecardSectionBase(BaseModel):
    """Base schema for scorecard section."""
    code: str = Field(..., max_length=20)
    name: str = Field(..., max_length=255)
    description: Optional[str] = None
    sort_order: int = 0
    is_active: bool = True


class ScorecardSectionCreate(ScorecardSectionBase):
    """Schema for creating a scorecard section."""
    pass


class ScorecardSectionUpdate(BaseModel):
    """Schema for updating a scorecard section."""
    name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class ScorecardSectionResponse(ScorecardSectionBase):
    """Response schema for scorecard section."""
    model_config = ConfigDict(from_attributes=True)

    section_id: int
    created_at: datetime
    updated_at: datetime


class ScorecardCriterionBase(BaseModel):
    """Base schema for scorecard criterion."""
    code: str = Field(..., max_length=20)
    section_id: int
    name: str = Field(..., max_length=255)
    description_prompt: Optional[str] = None
    comments_prompt: Optional[str] = None
    include_in_summary: bool = True
    allow_zero: bool = True
    weight: float = 1.0
    sort_order: int = 0
    is_active: bool = True


class ScorecardCriterionCreate(ScorecardCriterionBase):
    """Schema for creating a scorecard criterion."""
    pass


class ScorecardCriterionUpdate(BaseModel):
    """Schema for updating a scorecard criterion."""
    name: Optional[str] = Field(None, max_length=255)
    description_prompt: Optional[str] = None
    comments_prompt: Optional[str] = None
    include_in_summary: Optional[bool] = None
    allow_zero: Optional[bool] = None
    weight: Optional[float] = None
    sort_order: Optional[int] = None
    is_active: Optional[bool] = None


class ScorecardCriterionResponse(ScorecardCriterionBase):
    """Response schema for scorecard criterion."""
    model_config = ConfigDict(from_attributes=True)

    criterion_id: int
    created_at: datetime
    updated_at: datetime


class ScorecardSectionWithCriteria(ScorecardSectionResponse):
    """Section response with nested criteria."""
    criteria: List[ScorecardCriterionResponse] = []


class ScorecardConfigResponse(BaseModel):
    """Full scorecard configuration response."""
    sections: List[ScorecardSectionWithCriteria] = []


# ============================================================================
# Criterion Rating Schemas (Input/Output)
# ============================================================================

class CriterionRatingInput(BaseModel):
    """Input schema for a single criterion rating."""
    criterion_code: str = Field(..., max_length=20)
    rating: Optional[str] = Field(None, pattern=RATING_PATTERN)
    description: Optional[str] = None
    comments: Optional[str] = None


class ScorecardRatingsCreate(BaseModel):
    """Schema for creating/updating all scorecard ratings at once."""
    ratings: List[CriterionRatingInput] = Field(..., min_length=0)


class CriterionRatingUpdate(BaseModel):
    """Schema for updating a single criterion rating."""
    rating: Optional[str] = Field(None, pattern=RATING_PATTERN)
    description: Optional[str] = None
    comments: Optional[str] = None


class OverallNarrativeUpdate(BaseModel):
    """Schema for updating the overall assessment narrative."""
    overall_assessment_narrative: Optional[str] = Field(
        None,
        description="Free-text narrative for overall scorecard assessment"
    )


class CriterionRatingResponse(BaseModel):
    """Response schema for a criterion rating."""
    model_config = ConfigDict(from_attributes=True)

    rating_id: int
    request_id: int
    criterion_code: str
    rating: Optional[str] = None
    description: Optional[str] = None
    comments: Optional[str] = None
    created_at: datetime
    updated_at: datetime


# ============================================================================
# Computed Result Schemas
# ============================================================================

class CriterionDetailResponse(BaseModel):
    """Detailed criterion information with rating and score."""
    criterion_code: str
    criterion_name: str
    section_code: str
    rating: Optional[str] = None
    numeric_score: int
    description: Optional[str] = None
    comments: Optional[str] = None


class SectionSummaryResponse(BaseModel):
    """Section summary with computed score."""
    section_code: str
    section_name: str
    description: Optional[str] = None
    criteria_count: int
    rated_count: int
    unrated_count: int
    numeric_score: int
    rating: Optional[str] = None


class OverallAssessmentResponse(BaseModel):
    """Overall assessment with computed score."""
    numeric_score: int
    rating: Optional[str] = None
    sections_count: int
    rated_sections_count: int
    overall_assessment_narrative: Optional[str] = None


class ScorecardResultResponse(BaseModel):
    """Full scorecard result response."""
    model_config = ConfigDict(from_attributes=True)

    result_id: int
    request_id: int
    overall_numeric_score: Optional[int] = None
    overall_rating: Optional[str] = None
    overall_assessment_narrative: Optional[str] = None
    section_summaries: Optional[dict] = None
    config_snapshot: Optional[dict] = None
    computed_at: datetime


class ConfigCriterionResponse(BaseModel):
    """Criterion config for frontend rendering."""
    code: str
    name: str
    description_prompt: Optional[str] = None
    comments_prompt: Optional[str] = None
    allow_zero: bool = True
    weight: float = 1.0
    sort_order: int = 0


class ConfigSectionResponse(BaseModel):
    """Section config with nested criteria for frontend rendering."""
    code: str
    name: str
    description: Optional[str] = None
    sort_order: int = 0
    criteria: List[ConfigCriterionResponse] = []


class ScorecardFullResponse(BaseModel):
    """Complete scorecard response with all details."""
    request_id: int
    criteria_details: List[CriterionDetailResponse]
    section_summaries: List[SectionSummaryResponse]
    overall_assessment: OverallAssessmentResponse
    computed_at: datetime
    config_sections: List[ConfigSectionResponse] = []


# ============================================================================
# Legacy/Simple Config Response (for loading from JSON)
# ============================================================================

class SimpleSectionConfig(BaseModel):
    """Simple section config (matching SCORE_CRITERIA.json format)."""
    code: str
    name: str


class SimpleCriterionConfig(BaseModel):
    """Simple criterion config (matching SCORE_CRITERIA.json format)."""
    code: str
    section: str
    name: str
    description_prompt: Optional[str] = None
    comments_prompt: Optional[str] = None
    include_in_summary: bool = True
    allow_zero: bool = True
    weight: float = 1.0


class ScorecardJsonConfig(BaseModel):
    """Schema for SCORE_CRITERIA.json format."""
    sections: List[SimpleSectionConfig]
    criteria: List[SimpleCriterionConfig]


# ============================================================================
# Configuration Versioning Schemas
# ============================================================================

class ScorecardSectionSnapshotResponse(BaseModel):
    """Response schema for a section snapshot within a config version."""
    model_config = ConfigDict(from_attributes=True)

    snapshot_id: int
    code: str
    name: str
    description: Optional[str] = None
    sort_order: int
    is_active: bool


class ScorecardCriterionSnapshotResponse(BaseModel):
    """Response schema for a criterion snapshot within a config version."""
    model_config = ConfigDict(from_attributes=True)

    snapshot_id: int
    section_code: str
    code: str
    name: str
    description_prompt: Optional[str] = None
    comments_prompt: Optional[str] = None
    include_in_summary: bool
    allow_zero: bool
    weight: float
    sort_order: int
    is_active: bool


class ScorecardConfigVersionResponse(BaseModel):
    """Response schema for scorecard configuration version (summary)."""
    model_config = ConfigDict(from_attributes=True)

    version_id: int
    version_number: int
    version_name: Optional[str] = None
    description: Optional[str] = None
    published_by_name: Optional[str] = None
    published_at: datetime
    is_active: bool
    sections_count: int
    criteria_count: int
    scorecards_count: int
    created_at: datetime
    # True if config changed since this version was published
    has_unpublished_changes: bool = False


class ScorecardConfigVersionDetailResponse(ScorecardConfigVersionResponse):
    """Detailed response schema including all snapshots."""
    section_snapshots: List[ScorecardSectionSnapshotResponse] = []
    criterion_snapshots: List[ScorecardCriterionSnapshotResponse] = []


class PublishScorecardVersionRequest(BaseModel):
    """Request schema for publishing a new scorecard configuration version."""
    version_name: Optional[str] = Field(
        None, max_length=200, description="Optional display name for this version")
    description: Optional[str] = Field(
        None, description="Changelog or notes for this version")
